File: Editor/Tools/GetEditorStateTool.cs
using System;
using System.Reflection;
using UnityEditor;
using UnityEngine;
using UnityEngine.SceneManagement;
using Newtonsoft.Json.Linq;
using McpUnity.Unity;
using McpUnity.Utils;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool for checking the Unity Editor state, including modal dialogs, open windows,
    /// compilation/update status, and unsaved scene modifications.
    /// </summary>
    public class GetEditorStateTool : McpToolBase
    {
        public GetEditorStateTool()
        {
            Name = "get_editor_state";
            Description = "Gets the current Unity Editor state, including modal dialogs, focused/open windows, compilation/play status, and unsaved scene changes";
        }

        public override JObject Execute(JObject parameters)
        {
            try
            {
                var openWindows = new JArray();
                var modalWindows = new JArray();
                bool hasModalWindow = false;

                // Find all open EditorWindows
                var windows = UnityEngine.Resources.FindObjectsOfTypeAll<EditorWindow>();
                foreach (var window in windows)
                {
                    if (window == null) continue;

                    string typeName = window.GetType().Name;
                    string fullTypeName = window.GetType().FullName;
                    string title = window.titleContent != null ? window.titleContent.text : typeName;

                    bool isModalOrDialog = false;

                    // 1. Check window type name conventions
                    if (typeName.IndexOf("Modal", StringComparison.OrdinalIgnoreCase) >= 0 ||
                        typeName.IndexOf("Dialog", StringComparison.OrdinalIgnoreCase) >= 0 ||
                        typeName.IndexOf("Popup", StringComparison.OrdinalIgnoreCase) >= 0 ||
                        typeName.IndexOf("SaveModified", StringComparison.OrdinalIgnoreCase) >= 0 ||
                        typeName.IndexOf("Prompt", StringComparison.OrdinalIgnoreCase) >= 0 ||
                        typeName.IndexOf("Wizard", StringComparison.OrdinalIgnoreCase) >= 0 ||
                        typeName.IndexOf("Confirmation", StringComparison.OrdinalIgnoreCase) >= 0)
                    {
                        isModalOrDialog = true;
                    }

                    // 2. Check internal ContainerWindow showMode if accessible
                    try
                    {
                        var parentProp = typeof(EditorWindow).GetProperty("parent", BindingFlags.Instance | BindingFlags.NonPublic);
                        var hostView = parentProp?.GetValue(window);
                        if (hostView != null)
                        {
                            var windowProp = hostView.GetType().GetProperty("window", BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic);
                            var containerWindow = windowProp?.GetValue(hostView);
                            if (containerWindow != null)
                            {
                                var showModeProp = containerWindow.GetType().GetProperty("showMode", BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic);
                                if (showModeProp != null)
                                {
                                    object modeVal = showModeProp.GetValue(containerWindow);
                                    string modeStr = modeVal != null ? modeVal.ToString() : "";
                                    int modeInt = modeVal != null ? Convert.ToInt32(modeVal) : 0;
                                    // 2 = Utility, 6 = ModalUtility in Unity ContainerWindow.ShowMode
                                    if (modeInt == 2 || modeInt == 6 || modeStr.IndexOf("Modal", StringComparison.OrdinalIgnoreCase) >= 0)
                                    {
                                        isModalOrDialog = true;
                                    }
                                }
                            }
                        }
                    }
                    catch { }

                    var winJson = new JObject
                    {
                        ["title"] = title,
                        ["typeName"] = typeName,
                        ["fullTypeName"] = fullTypeName,
                        ["hasFocus"] = window.hasFocus,
                        ["maximized"] = window.maximized,
                        ["docked"] = window.docked,
                        ["isModalOrDialog"] = isModalOrDialog
                    };

                    openWindows.Add(winJson);
                    if (isModalOrDialog)
                    {
                        modalWindows.Add(winJson);
                        hasModalWindow = true;
                    }
                }

                // Check unsaved / dirty scenes
                var unsavedScenes = new JArray();
                int sceneCount = SceneManager.sceneCount;
                for (int i = 0; i < sceneCount; i++)
                {
                    Scene scene = SceneManager.GetSceneAt(i);
                    if (scene.isDirty || string.IsNullOrEmpty(scene.path))
                    {
                        unsavedScenes.Add(new JObject
                        {
                            ["name"] = scene.name,
                            ["path"] = scene.path,
                            ["isDirty"] = scene.isDirty,
                            ["isUntitled"] = string.IsNullOrEmpty(scene.path),
                            ["isLoaded"] = scene.isLoaded
                        });
                    }
                }

                EditorWindow focusedWindow = EditorWindow.focusedWindow;
                string focusedTitle = focusedWindow != null
                    ? (focusedWindow.titleContent != null ? focusedWindow.titleContent.text : focusedWindow.GetType().Name)
                    : null;

                var result = new JObject
                {
                    ["success"] = true,
                    ["type"] = "text",
                    ["message"] = hasModalWindow 
                        ? $"Editor has {modalWindows.Count} modal/dialog window(s) open"
                        : "Editor state retrieved successfully (no modal windows open)",
                    ["hasModalWindow"] = hasModalWindow,
                    ["modalWindowCount"] = modalWindows.Count,
                    ["modalWindows"] = modalWindows,
                    ["focusedWindow"] = focusedTitle,
                    ["openWindowCount"] = openWindows.Count,
                    ["openWindows"] = openWindows,
                    ["isCompiling"] = EditorApplication.isCompiling,
                    ["isUpdating"] = EditorApplication.isUpdating,
                    ["isPlaying"] = EditorApplication.isPlaying,
                    ["isPaused"] = EditorApplication.isPaused,
                    ["hasUnsavedChanges"] = unsavedScenes.Count > 0,
                    ["unsavedSceneCount"] = unsavedScenes.Count,
                    ["unsavedScenes"] = unsavedScenes
                };

                return result;
            }
            catch (Exception ex)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"Error getting editor state: {ex.Message}",
                    "editor_state_error"
                );
            }
        }
    }
}

File: Server~/src/tools/getEditorStateTool.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnity } from '../unity/mcpUnity.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import * as z from 'zod';
import { Logger } from '../utils/logger.js';

const toolName = 'get_editor_state';
const toolDescription = 'Gets the current Unity Editor state, including modal dialogs, focused/open windows, compilation/play status, and unsaved scene changes.';

const paramsSchema = z.object({});

export function registerGetEditorStateTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${toolName}`);

  server.tool(
    toolName,
    toolDescription,
    paramsSchema.shape,
    async (params: any) => {
      try {
        logger.info(`Executing tool: ${toolName}`, params);
        const result = await toolHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${toolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${toolName}`, error);
        throw error;
      }
    }
  );
}

async function toolHandler(mcpUnity: McpUnity, params: any) {
  const response = await mcpUnity.sendRequest({
    method: toolName,
    params
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to get editor state'
    );
  }

  let text = `Editor State:\n`;
  text += `  Has Modal Window: ${response.hasModalWindow}\n`;
  if (response.hasModalWindow && Array.isArray(response.modalWindows)) {
    text += `  Modal Windows (${response.modalWindowCount}):\n`;
    for (const win of response.modalWindows) {
      text += `    - "${win.title}" (${win.typeName})\n`;
    }
  }
  text += `  Focused Window: ${response.focusedWindow || '(none)'}\n`;
  text += `  Is Compiling: ${response.isCompiling}\n`;
  text += `  Is Updating: ${response.isUpdating}\n`;
  text += `  Is Playing: ${response.isPlaying}\n`;
  text += `  Is Paused: ${response.isPaused}\n`;
  text += `  Has Unsaved Changes: ${response.hasUnsavedChanges}\n`;
  if (response.hasUnsavedChanges && Array.isArray(response.unsavedScenes)) {
    text += `  Unsaved Scenes (${response.unsavedSceneCount}):\n`;
    for (const sc of response.unsavedScenes) {
      text += `    - "${sc.name}": ${sc.path || '(untitled)'}\n`;
    }
  }

  return {
    content: [{
      type: response.type || 'text',
      text
    }],
    structuredContent: {
      hasModalWindow: response.hasModalWindow,
      modalWindowCount: response.modalWindowCount,
      modalWindows: response.modalWindows,
      focusedWindow: response.focusedWindow,
      openWindowCount: response.openWindowCount,
      openWindows: response.openWindows,
      isCompiling: response.isCompiling,
      isUpdating: response.isUpdating,
      isPlaying: response.isPlaying,
      isPaused: response.isPaused,
      hasUnsavedChanges: response.hasUnsavedChanges,
      unsavedSceneCount: response.unsavedSceneCount,
      unsavedScenes: response.unsavedScenes
    }
  };
}
